### ISSUE TITLE
Run playwright in CI

### ISSUE THREAD

**Issue by fkling** [sourcegraph/sourcegraph#60881]

**Comment by bahrmichael** [comment]
`playwright install` (with or without extra args) seems blocked, because of permissions.

https://buildkite.com/sourcegraph/sourcegraph/builds/264611#018e2d93-f5a9-40c3-b8b8-b851b9567012

```
Installing dependencies...
E: setgroups 65534 failed - setgroups (1: Operation not permitted)
```

**Comment by bahrmichael** [comment]
No progress on hermetic install either.

## With hermetic env var

https://buildkite.com/sourcegraph/sourcegraph/builds/264631#018e2dc5-4475-4a6c-a531-ccd0266d934f

```
================================================================================
(13:48:31) FAIL: //client/web:playwright_install (see /root/.cache/bazel/_bazel_root/157e7a032b8afdaab0b3a3266b4b5f43/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/web/playwright_install/test.log)
FAILED: //client/web:playwright_install (Summary)
      /root/.cache/bazel/_bazel_root/157e7a032b8afdaab0b3a3266b4b5f43/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/web/playwright_install/test.log
      /root/.cache/bazel/_bazel_root/157e7a032b8afdaab0b3a3266b4b5f43/execroot/__main__/bazel-out/k8-fastbuild/testlogs/client/web/playwright_install/test_attempts/attempt_1.log
(13:48:31) INFO: From Testing //client/web:playwright_install:
==================== Test output for //client/web:playwright_install:
Installing dependencies...
E: setgroups 65534 failed - setgroups (1: Operation not permitted)
E: setegid 65534 failed - setegid (22: Invalid argument)
E: seteuid 100 failed - seteuid (22: Invalid argument)
```

## Without `--with-deps`:

https://buildkite.com/sourcegraph/sourcegraph/builds/264634#018e2dcd-f863-44c4-acaa-c4b451d14a7f

```
==================== Test output for //client/web:playwright_install:
Downloading Chromium 123.0.6312.4 (playwright build v1105) from https://playwright.azureedge.net/builds/chromium/1105/chromium-linux.zip
Error: getaddrinfo EAI_AGAIN playwright.azureedge.net
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:internal/dns/promises:95:17) {
  errno: -3001,
  code: 'EAI_AGAIN',
  syscall: 'getaddrinfo',
  hostname: 'playwright.azureedge.net'
}
Downloading Chromium 123.0.6312.4 (playwright build v1105) from https://playwright-akamai.azureedge.net/builds/chromium/1105/chromium-linux.zip
Error: getaddrinfo EAI_AGAIN playwright-akamai.azureedge.net
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:internal/dns/promises:95:17) {
  errno: -3001,
  code: 'EAI_AGAIN',
  syscall: 'getaddrinfo',
  hostname: 'playwright-akamai.azureedge.net'
}
Downloading Chromium 123.0.6312.4 (playwright build v1105) from https://playwright-verizon.azureedge.net/builds/chromium/1105/chromium-linux.zip
Error: getaddrinfo EAI_AGAIN playwright-verizon.azureedge.net
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:internal/dns/promises:95:17) {
  errno: -3001,
  code: 'EAI_AGAIN',
  syscall: 'getaddrinfo',
  hostname: 'playwright-verizon.azureedge.net'
}
Failed to install browsers
Error: Failed to download Chromium 123.0.6312.4 (playwright build v1105), caused by
Error: Download failure, code=1
    at ChildProcess.<anonymous> (/tmp/bazel-working-directory/__main__/bazel-out/k8-fastbuild/bin/client/web/playwright_install.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright-core@1.42.1/node_modules/playwright-core/lib/server/registry/browserFetcher.js:91:16)
    at ChildProcess.emit (node:events:514:28)
    at ChildProcess._handle.onexit (node:internal/child_process:294:12)
```

## With alternative but unrecommended approach

<img width="1016" alt="Screenshot 2024-03-11 at 15 15 43" src="https://github.com/sourcegraph/sourcegraph/assets/1830132/49437f68-de6b-437f-a932-56e17e1544ba">

Fails locally (if I npx uninstall playwright before) because it fails to load the new dependencies during Bazel runtime. It works fine when just running playwright or the npm scripts.

```
Error: browserType.launch: Executable doesn't exist at /private/var/tmp/_bazel_michael/680fb57cd51801cfe03bf19f9d7a0d3e/sandbox/darwin-sandbox/9132/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/client/web/playwright_test.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright-core@1.40.1/node_modules/playwright-core/.local-browsers/chromium-1091/chrome-mac/Chromium.app/Contents/MacOS/Chromium
```

**Comment by bahrmichael** [comment]
Another idea:

- instead of npm browser dependencies use http_archive to download a binary into a well-known path and pass that as the browser path env
- figure out where the binary from the browser dependencies is and pass that with the browser path env
- talk to James about Bazel and maybe running docker containers inside of it

**Comment by bahrmichael** [comment]
Main difference between puppeteer and playwright: Puppeteer downloads the browser during `npm install`; playwright has its own CDN and expectations on the folder structure where the apps live. I feel like mixing the two would be a bad idea.

**Comment by bahrmichael** [comment]
I'm stuck with a Bazel problem/understanding issues. After demonstrating that we can run playwright in CI (after installing the browser) and test arbitrary websites, I'm now trying to make this setup work with a preview server from the webapp rewrite. The main difference here is that there's no webserver running somewhere, but we also need to start and use the preview webserver.

The current BAZEL.build contains a call to vite to build a production build, which should then be used for the preview server.

https://github.com/sourcegraph/sourcegraph/blob/99d6bbabebb4c4f60472d4b209b9efd339d81480/client/web-sveltekit/BUILD.bazel#L112-L126

The current playwright config however builds and starts the preview server again:

https://github.com/sourcegraph/sourcegraph/blob/99d6bbabebb4c4f60472d4b209b9efd339d81480/client/web-sveltekit/playwright.config.ts#L8

My goal is that we can run playwright locally as well as on CI.

With that in mind I see a couple approaches we could take:

1. Keep the playwright.config.ts as is, and don't let the build and test targets depend on each other.
2. Create dedicated build, preview, and test targets in bazel, which depend on one another.
3. A mix of both.

WIP

---

The goal is to use two binaries: vite for running the preview server and playwright for running the tests. They need to run at the same time.

My preferred approach would be to use one bazel target to create a production build (with `vite build`) and then use another target to run playwright. The latter currently also spins up the preview server (with `vite preview`) based on the build target's output. I'm open to using another intermediate task to start the webserver and use that long running process in the playwright test target, but I don't know how I would do that yet.

Here's what I'm currently basing my experiments off:

playwright.config.ts
```typescript
const config: PlaywrightTestConfig = {
    testMatch: '**/*.spec.ts',
    reporter: 'list',
    timeout: 5_000,
    webServer: {
        command: 'npm run build && npm run preview',
        port: PORT,
        reuseExistingServer: !process.env.CI,
    },
    use: {
        baseURL: `http://localhost:${PORT}`,
    },
    projects: [
        {
            name: 'chromium',
            use: { ...devices['Desktop Chrome'] },
        },
    ]
}
```

BAZEL.build
```starlark
vite_bin.vite(
    name = "web-sveltekit",
    srcs = SRCS + BUILD_DEPS + CONFIGS,
    args = [
        "build",
        "--force",
        "-c vite.config.ts",
    ],
    chdir = package_name(),
    env = {
        "BAZEL": "1",
    },
    out_dirs = ["build"],
    visibility = ["//client/web/dist:__pkg__"],
    # silent_on_success = False,
)

playwright_test_bin.playwright_test(
    name = "e2e_test",
    args = [
        "test",
        "--config ./playwright.config.ts",
    ],
    chdir = package_name(),
    data = glob(["**/*.spec.ts"]) + [
        "package.json",
        # ":playwright_install",  # removed this to speed up the feedback loop, will be brought back later
        "playwright.config.ts",
        ":web-sveltekit",
    ],
    env = {
        "BAZEL": "1",
    },
    local = True,
    log_level = "info",
    tags = [
        "no-sandbox",
        "requires-network",
    ],
)
```

This setup has a few issues:

1. `package.json` is used by both targets (to run the build and preview commands) and leads to a conflict: `ERROR: file 'client/web-sveltekit/package.json' is generated by these conflicting actions:
Label: //client/web-sveltekit:e2e_test, //client/web-sveltekit:web-sveltekit_copy_srcs_to_bin`
5. When I attempt to resolve that conflict, I either break the web-sveltekit target, or by using `copy_srcs_to_bin=False` on the e2e_test target I break that one as it can't find the `vitest` executable. I don't know if `vitest` is properly copied in (1).
6. 


Alternatives considered:

* Using `npm run build && npm run preview` as the playwright.config.ts command. This leaves us with the same problem that vite and the package.json is required for either of them.

**Comment by bahrmichael** [comment]
Good catch! I just added the commit https://github.com/sourcegraph/sourcegraph/pull/60897/commits/c905a5dbd773ffcc0601fac9a15680adc8b2574e which I forgot to create yesterday, and fixed the import there. However I'm still getting the same error.

The previous playwright prototype also used the same import, so there should be nothing wrong with it.

```
load("@npm//client/playwright:@playwright/test/package_json.bzl", playwright_test_bin = "bin")
```

I have not been able to find any other dependencies that pull in playwright and lead to conflicts.

**Comment by jamesmcnamara** [comment]
@bahrmichael What we've done in the past for this is just stub out any calls from the test driver to just return static files. So for example we'd [intercept all calls to the server URL](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@995a6553422df54f1bb99df2a4cf9f3e846b2fb3/-/blob/client/web/src/integration/context.ts?L67-78) and statically return the manifest file. Then in turn we'd [catch any asset file requests](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@995a6553422df54f1bb99df2a4cf9f3e846b2fb3/-/blob/client/shared/src/testing/integration/context.ts?L194-216) and serve them from the file system. 

The benefit is that the test process is greatly simplified; there aren't two parallel communicating processes that need to be brought up, managed and torn down, just a single process (which you already have) which resolves all of its own requests. Particularly in CI this makes life much easier, and the same approach can be run locally.

The drawbacks that I can think of are:
1. This can be quite surprising behavior. When people see browser tests people assume there is some amount of backend running (However this is already somewhat true if you are mocking out API responses)
2. If your web server is doing non-trivial routing logic, that must be duplicated or imported into your test and maintained in parallel (However if you're just trying to recreate a `vite preview` web server this shouldn't apply)

If you like this approach I can work on it tomorrow (not sure how long it will take because I don't know enough about how the svelte app assets bundling / manifest / requests differ from the existing react-app).

If for some reason we need to run `vite preview` unequivocally, then I'll have to do some more research. What we have today for our "true" e2e tests are [`server_integration_tests`](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@995a6553422df54f1bb99df2a4cf9f3e846b2fb3/-/blob/testing/defs.bzl?L1:5-1:28) like [e2e_test](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@995a6553422df54f1bb99df2a4cf9f3e846b2fb3/-/blob/testing/BUILD.bazel?L3-42) that actually start up a [docker image via a bash script](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@995a6553422df54f1bb99df2a4cf9f3e846b2fb3/-/blob/testing/e2e_test.sh), so not the most official or easily supervisable technique

**Comment by bahrmichael** [comment]
@jamesmcnamara Thank you very much for your input! I took some time to investigate the intercept approach you suggested, and while not able to make a test green while intercepting all calls I feel like this might be the better approach.

You can see my attempt in https://github.com/sourcegraph/sourcegraph/pull/60897/commits/059a1cb9cd00abe7c1364b9460ea9fcc5405e3a8, where I've set up most intercepts except for the indexHtml. I skipped that part because I wanted to get an understanding of polly and the server intercepts before jumping into our custom code and figuring out what needs to be copied/adapated.

If you want to pick up here and make the finishing touches on the index/manifest part, this section might be a good starting point to add new interceptors to: https://github.com/sourcegraph/sourcegraph/pull/60897/commits/059a1cb9cd00abe7c1364b9460ea9fcc5405e3a8#diff-fe440130504bd56a8be70fd3da04f0d03ac583cc4fd8134fa96f64b09120fd26R61-R67

I've seen the `_app/version.json` request running through my interceptor, so the setup seems to be working in general.

When we run `pnpm run build` we get a `build` folder that holds the assets we want to test against, and I found the `manifest.json` in the `.svelte-kit` folder. I'm not confident enough about manifest files to say that this is the right one.

**Comment by jamesmcnamara** [comment]
@bahrmichael I added a commit that makes all the tests green :) I removed Polly and just used playwright's native [network mocking](https://playwright.dev/docs/network). As far as the manifest, we're in luck that vite actually produces a valid index.html file with the appropriately generated entrypoints, so we don't have to worry about the manifest at all and can just resolve the requests straight to index.html.

**Comment by bahrmichael** [comment]
New insights!

I think that this is not actually caused by multiple versions of playwright, but by a combination of how playwright, bazel and node work. The culprit here is is a conflict between [how Playwright loads files](https://github.com/microsoft/playwright/blob/main/packages/playwright/src/common/globals.ts), and that Bazel works a lot with symlinks which leads to the same playwright module being loaded multiple times.

I stumbled upon [a playwright issue](https://github.com/microsoft/playwright/issues/9193) where they had a bug related to lower/upper casing of directories. I added the snippet from the issue to the globals file that Playwright uses during Bazel, and got confirmation that there is indeed a loading conflict.

```javascript
const s = Symbol.for('__playwright');
if (globalThis[s]) {
  throw new Error(`more than one instance of globals.js was loaded`);
} else {
  Reflect.defineProperty(globalThis, s, { value: true, configurable: true });
}
```

```
INFO: aspect_rules_js[js_test]: running /.../bazel-out/darwin_arm64-fastbuild/bin/client/web-sveltekit/e2e_test.sh.runfiles/__main__/client/web-sveltekit/e2e_test_node_bin/node --preserve-symlinks-main -- /.../bazel-out/darwin_arm64-fastbuild/bin/client/web-sveltekit/e2e_test.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright/cli.js test --config ./playwright.config.ts
Error: more than one instance of globals.js was loaded
    at Object.<anonymous> (/.../bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright/lib/common/globals.js:5:9)
```

There's also [this issue](https://github.com/microsoft/playwright/issues/13092) where people are seeing similar problems.

Some options we can explore to resolve this issue:

- Figure out what's wrong with Bazel symlinks
- Write a patch for playwright (might need a fork)
- Use the nodejs_binary (=> js_test with aspect) approach [mentioned here](https://github.com/microsoft/playwright/issues/9193#issuecomment-1048969266)

**Comment by bahrmichael** [comment]
I tried three variations of the approach to use js_test instead with an entrypoint.

## node_modules entry with playwright dependency

```
js_test(
    name = "e2e_test",
    args = [
        "test",
        "--config ./playwright.config.ts",
    ],
    chdir = package_name(),
    data = [
        ":test_files",
        ":web-sveltekit",
        "//client/web-sveltekit:node_modules/@playwright/test",
        "//client/web-sveltekit:node_modules/playwright",
    ],
    entry_point = "node_modules/playwright/cli.js",
    env = {
        "BAZEL": "1",
        "PW_RUNNER_DEBUG": "true",
    },
    local = True,
    log_level = "info",
    tags = [
        "no-sandbox",
        "requires-network",
    ],
)
```

Here we're telling js_test to call `node_modules/playwright/cli.js`. This however leads to a conflict between the data import and the entrypoint:

```
ERROR: One of the output paths 'bazel-out/darwin_arm64-fastbuild/bin/client/web-sveltekit/node_modules/playwright/cli.js' (belonging to //client/web-sveltekit:e2e_test) and 'bazel-out/darwin_arm64-fastbuild/bin/client/web-sveltekit/node_modules/playwright' (belonging to //client/web-sveltekit:node_modules/playwright) is a prefix of the other. These actions cannot be simultaneously present; please rename one of the output files or build just one of them
```

## node_modules entrypoint without dependency

```
js_test(
    name = "e2e_test",
    args = [
        "test",
        "--config ./playwright.config.ts",
    ],
    chdir = package_name(),
    data = [
        ":test_files",
        ":web-sveltekit",
        "//client/web-sveltekit:node_modules/@playwright/test",
        # "//client/web-sveltekit:node_modules/playwright",
    ],
    entry_point = "node_modules/playwright/cli.js",
    env = {
        "BAZEL": "1",
        "PW_RUNNER_DEBUG": "true",
    },
    local = True,
    log_level = "info",
    tags = [
        "no-sandbox",
        "requires-network",
    ],
)
```

This leads to an error, because Bazel only copies the entrypoint, but not the rest of the module that playwright needs:

```
Error: Cannot find module './lib/program'
Require stack:
- /private/var/tmp/_bazel_michael/680fb57cd51801cfe03bf19f9d7a0d3e/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/client/web-sveltekit/e2e_test.sh.runfiles/__main__/client/web-sveltekit/node_modules/playwright/cli.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
```

## Use a dedicated script to start the playwright process with playwright dependencies in data

Here's the Bazel step:

```
js_test(
    name = "e2e_test",
    args = [
        "test",
        "--config ./playwright.config.ts",
    ],
    chdir = package_name(),
    data = [
        # ":playwright_install",
        ":test_files",
        ":web-sveltekit",
        "//client/web-sveltekit:node_modules/@playwright/test",
        "//client/web-sveltekit:node_modules/playwright",
    ],
    entry_point = "playwright_runner.js",
    env = {
        "BAZEL": "1",
        "PW_RUNNER_DEBUG": "true",
    },
    local = True,
    log_level = "info",
    tags = [
        "no-sandbox",
        "requires-network",
    ],
)
```

And here's the playwright runner:

```javascript
import {spawnSync} from 'child_process';

const result = spawnSync('node node_modules/playwright/cli.js test --config playwright.config.ts', { encoding: 'utf8', shell: '/bin/bash' });
console.error('Execution Result: \n', result.stdout);
throw Error(JSON.stringify(result.stderr))
```

With this we get the "dependency" error again, so I think this one didn't help with fixing Bazel symlinks.

```
Error: Playwright Test did not expect test.describe() to be called here.
Most common reasons include:
- You are calling test.describe() in a configuration file.
- You are calling test.describe() in a file that is imported by the configuration file.
- You have two different versions of @playwright/test. This usually happens
  when one of the dependencies in your package.json depends on @playwright/test.
```

**Comment by jamesmcnamara** [comment]
@bahrmichael Alright I made some progress on this [here](https://github.com/sourcegraph/sourcegraph/pull/60897/commits/2bb9399367da9617e2a061161a95d2fad339574d). After a lot of digging it seems the problem comes from there actually being two copies of playwright somehow in the bundle. After cutting down a lot of the overlapping dependencies in the main `:web-sveltekit` target, I got some different errors.

The changes in that commit omits the dep on :web_sveltekit (so it won't work) but it's actually able to run all of our tests, but they just fail when they try to load `build/index.html` because that is an output of :web_sveltekit. So I feel like somewhere in that target we're pulling in another copy of playwright somehow. 

The other issue is that our test fixture calls `import.meta.url` to construct the path to the schema object which isn't available when this is run in commonjs land. So we'll have to either find a way to pre-compile that code as a module or replace that path with something else.

I'll keep looking at this tomorrow.

**Comment by bahrmichael** [comment]
@jamesmcnamara Very happy to see your progress! I can confirm that without importing `:web-sveltekit` into the `e2e_test` target I get the same errors that you describe.

I tried to dive deeper into this issue, and tried some variations (e.g. using playwright instead of `@playwright/test` in the build file or using `preserve_symlinks_main = False,` on the `e2e_test` target. I also renamed `:web-sveltekit` to `:build` locally to test if there are any shadowing issues between the package name and the target. No change in results though.

Another approach I took was to add a `console.trace` into the `globals.js` of playwright (aside from the script that causes an error if that fail is loaded twice).

```
$ head client/web-sveltekit/e2e_test.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright/lib/common/globals.js
"use strict";

const s = Symbol.for('__playwright');
if (globalThis[s]) {
  throw new Error(`more than one instance of globals.js was loaded`);
} else {
  console.trace() // This trace is new
  Reflect.defineProperty(globalThis, s, { value: true, configurable: true });
}
```

The trace inside the `globals.js` points to two files on the system, which should be the same based on their symlink:

```
bazel-out/darwin_arm64-fastbuild/bin/client/web-sveltekit/e2e_test.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright/lib/common/globals.js:7:11
```

```
bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright/lib/common/globals.js:7:11
```

Symlink:

```
$ cd bazel-out/darwin_arm64-fastbuild
$ ls -al client/web-sveltekit/e2e_test.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright    
lrwxr-xr-x@ 1 michael  wheel  191 Mar 29 17:25 client/web-sveltekit/e2e_test.sh.runfiles/__main__/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright -> /private/var/tmp/_bazel_michael/680fb57cd51801cfe03bf19f9d7a0d3e/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright
```

After two traces appear, we then see an error from this file:

```
bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/playwright@1.42.1/node_modules/playwright/lib/common/globals.js:5:9
```

I find it interesting that two traces can appear before an error is raised. I was expecting only one trace, but then I'm not sure how well node deals with symlinks.

---

If this task is turning out too large, I'll be happy to recap things into a new issue :)

**Comment by jamesmcnamara** [comment]
Little update: I've zeroed in on the exact line that is causing the issue (in brief if the test suites are loaded as ESM it's reinitializing playwright in a new module context, but if it loads it as CommonJS, references to playwright are resolved correctly). I'll keep looking at this tomorrow

**Comment by bahrmichael** [comment]
@jamesmcnamara Just wanted to check in if you've been able to make more progress. Since I'm not able to make progress myself I want to be supportive with whatever approach you think is best :) If the challenge of ESM vs. CommonJS is turning out to difficult, we could also revisit [the "true" web server approach](https://github.com/sourcegraph/sourcegraph/issues/60881#issuecomment-2019115005). Furthermore, if this is turning out too large of an effort to do on the side, I'll be happy to funnel this into something more official. Let me know what you think is the best path forward at this time.

**Comment by bahrmichael** [comment]
I'm continuing to track this ticket. My current status is that things have been set up so that we can work with Aspect.
